doc/developer/design/20230512_mir_cost_model.md

```diff
@@ -77,22 +77,24 @@ Traditional databases track cardinality data and statistics about various column
 
 ### Symbolic costs
 
-A symbolic cost is a polynomial over the features of the input relations. For example, a batch query that loops over relation `R` twice and then computes its cross product with relation `Q` will have cost `2|R| + |R|*|Q|`. A traditional database typically computes cost just in terms of overall batch query time, but our costs are multi-dimensional. In rough feasibility order:
+A symbolic cost is a polynomial over the features of the input relations. For example, a batch query that loops over relation `R` twice and then computes its cross product with relation `Q` will have cost `2|R| + |R|*|Q|`. A traditional database typically computes cost just in terms of overall batch query time, but our costs are multi-dimensional. In priority order:
 
-| Dimension          | Input features                                               | Outputs                         | Output type |
-| ------------------ | ------------------------------------------------------------ | ------------------------------- | ----------- |
-| Hydration time     | Source cardinality                                           | Time estimate                   |             |
-| Latency per update | Source cardinality, expected per-update delta on each source | Time estimate (per source)      |             |
-| Storage            | Source cardinality, existing indices, cardinality of groups  | Arrangements and size estimates |             |
-| Scalability        | Distribution of keys (e.g., cardinality of distinct keys)    | Data parallelism estimate       |             |
+| Dimension          | Input features                                               | Outputs                         | Output type                         |
+| ------------------ | ------------------------------------------------------------ | ------------------------------- | ----------------------------------- |
+| Memory             | Source cardinality, existing indices, cardinality of groups  | Arrangements and size estimates | (Vec<Arrangement>, Formula<Source>) |
+| Hydration time     | Source cardinality                                           | Time estimate                   | Formula<Source>                     |
+| Latency per update | Source cardinality, expected per-update delta on each source | Time estimate (per source)      | Map<Source, Formula<Source>>        |
+| Scalability        | Distribution of keys (e.g., cardinality of distinct keys)    | Data parallelism estimate       | Map<Source, Vec<Column>>            |
+
+In the output type, `Arrangement` refers to summary information about the arrangements necessary; `Formula<V>` refers to an algebraic expression with variables drawn from the type `V`.
+
+The *memory* dimension is meant to identify particularly expensive new arrangements: for example, a count query grouping by US state code (just 50, no big deal) vs. grouping by user ID (as big as your user table).
```

Comment:
In the context of the notation [from the above comment](https://github.com/MaterializeInc/materialize/pull/19260/files/2386a1525aeb349dca39371d4b3370daf26d2b84..d133fe56087b204385d4815d7730cc6da7bdc1d5#r1209316841), I think we might be able to not only use `cost(M)` internally for optimizations, but also use `EXPLAIN WITH(memory)` to annotate each sub-plan with its inferred `M` value and use this to teach users how to reason about the expected memory consumption of their queries.

cc @ggnall this is in line with point (1) from [your comment here](https://github.com/MaterializeInc/materialize/pull/19381#issuecomment-1559536505), and something that can be addressed only after changes in the `EXPLAIN` and/or dataflow visualizer infrastructure. Let me know if you would like to have this as a concrete outcome from this effort, so I can discuss with @uce how we would deliver this.